### PR TITLE
Complete color pipeline overhaul

### PR DESCRIPTION
**Pull request by dictoon**
The color pipeline is now configurable to use RGB colors or full spectra. The color pipeline mode ("spectrum mode") is stored in a thread-local static variable. All `DynamicSpectrum<>` objects now always have the same size within a thread. We no longer upgrade RGB colors to spectra or downgrade spectra to RGB colors. `DynamicSpectrum<>::resize()` is gone.

Spectra intent are now only use when converting between RGB colors and spectra, and are no longer stored. Consequently, the size of a 31-channel `DynamicSpectrum<>` is now exactly 128 bytes (31 floats + 1 float of padding).

The result of shading a pixel sample is now always in linear RGB and is represented by a `foundation::Color4f` object. The classes `ShadingFragment`, `ShadingFragmentStack` and `SpectrumStack` are gone.

Cornell Box-based test scenes and scenes from the "sun and sky" test suite have been switched to spectral mode; all other scenes are set to use RGB mode.

All test suite images were thoroughly checked, then updated.

### CODE REVIEW

File: src/appleseed/renderer/kernel/aov/aovaccumulator.h [hunk 1]
```diff
@@ -37,14 +37,16 @@
...
 
+// Standard headers.
+#include <cstddef>
+
 // Forward declarations.
-namespace renderer { class Camera; }
-namespace renderer { class ShadingComponents; }
-namespace renderer { class ShadingPoint; }
-namespace renderer { class ShadingResult; }
+namespace renderer  { class Camera; }
```
Comment (Narann) [inline src/appleseed/renderer/kernel/aov/aovaccumulator.h]: Two spaces? I would be interested to know the standard using this. Is it personal taste?

File: src/appleseed/renderer/kernel/lighting/pathtracer.h [hunk 2]
```diff
@@ -768,30 +768,33 @@ bool PathTracer<PathVisitor, VolumeVisitor, Adjoint>::march(
...
         const float s = sampling_context.next2<float>();
-        const size_t channel = foundation::truncate<size_t>(s * channel_count);
-        const bool extinction_is_null = (extinction_coef[channel] < 1.0e-6f);
+        const size_t channel = foundation::truncate<size_t>(s * Spectrum::size());
+        const bool extinction_is_null = extinction_coef[channel] < 1.0e-6f;
 
-        float distance_sample = 0.0f;
-        float distance_pdf = 0.0f;
-        if (!extinction_is_null)
+        // Sample distance.
```
Comment (Narann) [inline src/appleseed/renderer/kernel/lighting/pathtracer.h]: I suggest to remove the comment or improve it. The original comment position looks clearer IMHO.

File: src/appleseed/renderer/kernel/lighting/pathtracer.h [hunk 3]
```diff
@@ -768,30 +768,33 @@ bool PathTracer<PathVisitor, VolumeVisitor, Adjoint>::march(
...
+        const Spectrum& extinction_coef =
+            volume->extinction_coefficient(vertex.m_volume_data, volume_ray);
         
         sampling_context.split_in_place(1, 2);
 
         // Sample channel uniformly at random.
         const float s = sampling_context.next2<float>();
-        const size_t channel = foundation::truncate<size_t>(s * channel_count);
-        const bool extinction_is_null = (extinction_coef[channel] < 1.0e-6f);
+        const size_t channel = foundation::truncate<size_t>(s * Spectrum::size());
```
Comment (Narann) [inline src/appleseed/renderer/kernel/lighting/pathtracer.h]: Does it mean in rgb `Spectrum::size()` return `3`?

File: src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp [hunk 4]
```diff
@@ -131,7 +126,7 @@ void ShadingResultFrameBuffer::merge(
     const float* APPLESEED_RESTRICT source_ptr = source.pixel(source_x, source_y);
     float* APPLESEED_RESTRICT dest_ptr = pixel(dest_x, dest_y);
 
-    for (size_t i = 0; i < m_channel_count; ++i)
+    for (size_t i = 0, e = m_channel_count; i < e; ++i)
```
Comment (Narann) [inline src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp]: Why `e`? Why not `for (size_t i = 0; i <  m_channel_count; ++i)`?

File: src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp [hunk 5]
```diff
@@ -152,7 +147,7 @@ void ShadingResultFrameBuffer::develop_to_tile_premult_alpha(
             tile.set_pixel(x, y, color * rcp_weight);
             ptr += 4;
 
-            for (size_t i = 0; i < m_aov_count; ++i)
+            for (size_t i = 0, e = m_aov_count; i < e; ++i)
```
Comment (Narann) [inline src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp]: Same here: Why `e`, Why not `for (size_t i = 0; i < m_aov_count; ++i)`? Access?

File: src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp [hunk 6]
```diff
@@ -168,9 +163,9 @@ void ShadingResultFrameBuffer::develop_to_tile_straight_alpha(
 {
     const float* ptr = pixel(0);
 
-    for (size_t y = 0; y < m_height; ++y)
+    for (size_t y = 0, h = m_height; y < h; ++y)
     {
-        for (size_t x = 0; x < m_width; ++x)
+        for (size_t x = 0, w = m_width; x < w; ++x)
```
Comment (Narann) [inline src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp]: Mmh, you us this trick everywhere. Now I guess there is a reason.

File: src/appleseed/renderer/modeling/aov/depthaov.cpp [hunk 7]
```diff
@@ -80,8 +81,7 @@ namespace
 
         virtual void flush(ShadingResult& result) override
         {
-            result.m_aovs[m_index].m_color.set(m_depth);
-            result.m_aovs[m_index].m_alpha.set(1.0f);
+            result.m_aovs[m_index] = Color4f(m_depth, m_depth, m_depth, 1.0f);
```
Comment (Narann) [inline src/appleseed/renderer/modeling/aov/depthaov.cpp]: No `set()` method on `m_aovs` instead a copy? I don't know.

File: src/appleseed/renderer/modeling/aov/diffuseaov.cpp [hunk 8]
```diff
@@ -73,18 +75,18 @@ namespace
...
-            m_color = shading_components.m_diffuse;
+            m_color = shading_components.m_diffuse.to_rgb(g_std_lighting_conditions);
             m_color *= multiplier;
         }
 
         virtual void flush(ShadingResult& result) override
         {
-            result.m_aovs[m_index].m_color = m_color;
-            result.m_aovs[m_index].m_alpha = result.m_main.m_alpha;
+            result.m_aovs[m_index].rgb() = m_color;
```
Comment (Narann) [inline src/appleseed/renderer/modeling/aov/diffuseaov.cpp]: An assignment on a method `.rgb()`? What is that?

File: src/appleseed/renderer/utility/dynamicspectrum.h [hunk 9]
```diff
@@ -833,32 +631,16 @@ APPLESEED_FORCE_INLINE DynamicSpectrum<float, 31>& operator+=(DynamicSpectrum<fl
...
-    else
-    {
-        DynamicSpectrum<T, N> up_rhs;
-        DynamicSpectrum<T, N>::upgrade(rhs, up_rhs);
-
-        for (size_t i = 0; i < N; ++i)
-            lhs[i] -= up_rhs[i];
-    }
+    for (size_t i = 0; i < DynamicSpectrum<T, N>::size(); ++i)
+        lhs[i] -= rhs[i];
```
Comment (Narann) [inline src/appleseed/renderer/utility/dynamicspectrum.h]: No `_mm_sub_ps` version for 31?

File: sandbox/tests/test scenes/alpha/01 - opaque red over opaque blue.appleseed [hunk 10]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- File generated by appleseed.dll version 1.7.0-beta (SSE 4.2). -->
+<!-- File generated by appleseed.dll version 1.8.0-beta (SSE 2). -->
```
Comment (dictoon) [inline sandbox/tests/test scenes/alpha/01 - opaque red over opaque blue.appleseed]: It just so happens that my current build (which I used to automatically update all the test scenes) doesn't have SSE 4.2 enabled. Not a big deal.

File: src/appleseed/renderer/kernel/aov/aovaccumulator.h (hunk 1 above)
Comment (dictoon) [inline src/appleseed/renderer/kernel/aov/aovaccumulator.h]: There's no standard really, I just press tab to align the opening brace on a tab stop...

File: src/appleseed/renderer/kernel/lighting/pathtracer.h (hunk 3 above)
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/pathtracer.h]: Yes, correct.

File: src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp (hunk 4 above)
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/shadingresultframebuffer.cpp]: This is to help the reader and especially the compiler: this construction ensures that the loop's upper bound (`m_channel_count`) doesn't change during iteration, so the compiler doesn't have to reload it at each iteration.

File: src/appleseed/renderer/modeling/aov/depthaov.cpp (hunk 7 above)
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/depthaov.cpp]: Yeah, it bothers me a little sometimes too.

`m_avos[m_index]` is a `foundation::Color4f`, which has a `set()` method to set all components to the same value, but if you want to set all components individually, right now there's no other choice than making a copy. I'm pretty sure the compiler will elide this copy, but having a `set(r, g, b, a)` method would be another choice.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 11]
```diff
@@ -703,15 +703,15 @@ namespace
         {
             const string widget_base_key = prefix + ".bounces.";
 
-            QSpinBox* max_bounces = create_integer_input(widget_base_key + "max_bounces", 0, 10000, 1);
+            QSpinBox* max_bounces = create_integer_input(widget_base_key + "max_bounces", 0, 100, 1);
```
Comment (est77) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Would it make sense to make this a constant and reference it everywhere.
Would be useful if we need to adjust the max bounces limit in the future.

File: src/appleseed/renderer/kernel/aov/aovaccumulator.cpp [hunk 12]
```diff
@@ -89,16 +88,12 @@ BeautyAOVAccumulator::BeautyAOVAccumulator()
 
 void BeautyAOVAccumulator::set(const Spectrum& value)
 {
-    m_color_space = value.is_rgb() ? ColorSpaceLinearRGB : ColorSpaceSpectral;
-    m_color = value;
+    m_color = value.to_rgb(g_std_lighting_conditions);
```
Comment (est77) [inline src/appleseed/renderer/kernel/aov/aovaccumulator.cpp]: Can we make the lighting conditions a thread local variable too?
In our color management PR, lighting conditions are selected as part of the working color space selection and g_std_lighting_conditions does not exist anymore.

File: src/appleseed/renderer/kernel/lighting/pt/ptlightingengine.cpp [hunk 13]
```diff
@@ -1047,14 +1041,10 @@ namespace
...
-                Spectrum precomputed_mis_weights;
-                if (channel_count == scattering_coef.size())
-                    precomputed_mis_weights = scattering_coef;
-                else
-                    Spectrum::upgrade(scattering_coef, precomputed_mis_weights);
+                Spectrum precomputed_mis_weights = scattering_coef;
                 precomputed_mis_weights *= vertex.m_throughput;
                 float max_density = 0.0f;
-                for (size_t i = 0; i < channel_count; ++i)
+                for (size_t i = 0; i < Spectrum::size(); ++i)
```
Comment (est77) [inline src/appleseed/renderer/kernel/lighting/pt/ptlightingengine.cpp]: e = Spectrum::size(); i < e; ++i) ?

File: src/appleseed/renderer/kernel/lighting/sppm/sppmparameters.cpp [hunk 14]
```diff
@@ -96,7 +96,8 @@ namespace
 }
 
 SPPMParameters::SPPMParameters(const ParamArray& params)
-  : m_sampling_mode(get_sampling_context_mode(params))
+  : m_spectrum_mode(get_spectrum_mode(params))
```
Comment (est77) [inline src/appleseed/renderer/kernel/lighting/sppm/sppmparameters.cpp]: Right now, we are collecting all color related parameters in the Frame entity.
Would it make sense to move spectrum_mode there too?

File: src/appleseed/renderer/modeling/aov/diffuseaov.cpp (hunk 8 above)
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/diffuseaov.cpp]: `rgb()` returns the first three components (R, G, B) as a `foundation::Color3f` reference so assignment works just fine.

File: src/appleseed/renderer/utility/dynamicspectrum.h (hunk 9 above)
Comment (dictoon) [inline src/appleseed/renderer/utility/dynamicspectrum.h]: Most arithmetic operations are not used (at least not in performance-sensitive contexts) so they haven't been optimized yet.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp (hunk 11 above)
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: I thought about it but was lazy. But you're right, I'll do it.

File: src/appleseed/renderer/kernel/lighting/pt/ptlightingengine.cpp (hunk 13 above)
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/pt/ptlightingengine.cpp]: I was doing that at first, then reverted to this simpler form... But after more thinking, it appears that maybe you're right and that I should change it everywhere. I'll check if there's any difference in the generated code.

File: src/appleseed/renderer/modeling/aov/diffuseaov.cpp (hunk 8 above)
Comment (Narann) [inline src/appleseed/renderer/modeling/aov/diffuseaov.cpp]: Rhooo that's awesome! :D

File: src/appleseed/renderer/utility/dynamicspectrum.h (hunk 9 above)
Comment (Narann) [inline src/appleseed/renderer/utility/dynamicspectrum.h]: Ok, thanks!

**Pull request merged**